### PR TITLE
add: post on deploying tf vision models with tf serving.

### PR DESCRIPTION
**Pull request by sayakpaul**
@osanseviero FYI. 

I wrote the draft in a Google Doc and I have Grammarly incorporated into it. I have made a pass through Grammarly for catching any typos or grammatical errors. But happy to correct if there's still any.

### CODE REVIEW

File: tf-serving-vision.md [hunk 1]
```diff
@@ -0,0 +1,490 @@
+---
+title: Deploying TensorFlow Vision Models in Hugging Face with TF Serving
```
Comment (sayakpaul) [inline tf-serving-vision.md:2]: Not sure how to generate a thumbnail. Any resource will be helpful. I can add it after that.

File: tf-serving-vision.md [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+---
+
+<h1>
+  Deploying TensorFlow Vision Models in Hugging Face with TF Serving
+</h1>
+
+<div class="blog-metadata">
+    <small>Published July 19, 2022.</small>
```
Comment (sayakpaul) [inline tf-serving-vision.md:10]: Of course, will change this to the merge date after the PR is approved?

File: tf-serving-vision.md [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+TF Serving's request payload format specification for the REST endpoint
+is available [here](https://www.tensorflow.org/tfx/serving/api_rest#request_format_2).
+Within the `instances` we can pass multiple encoded images. This kind
+of endpoints are meant to be consumed for online prediction scenarios.
+For inputs having more than a single data point, you would to want to
+[enable batching](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/README.md)
+to get performance optimization benefits.
+
+Now we can call the API:
+
```
Comment (osanseviero) [inline tf-serving-vision.md:368]:
```suggestion
TF Serving's request payload format specification for the REST endpoint
is available [here](https://www.tensorflow.org/tfx/serving/api_rest#request_format_2).
Within the `instances`, you can pass multiple encoded images. This kind
of endpoints are meant to be consumed for online prediction scenarios.
For inputs having more than a single data point, you would to want to
[enable batching](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/batching/README.md)
to get performance optimization benefits.

Now you can call the API:

```

File: tf-serving-vision.md [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+print(json.loads(json_response.text))
+# {'predictions': [{'label': 'Egyptian cat', 'confidence': 0.896659195}]}
+
+```
+
+Our REST API is -
+`http://localhost:8501/v1/models/vit:predict` following the specification from
+[here](https://www.tensorflow.org/tfx/serving/api_rest#predict_api). By default,
+this always picks up the latest version of the model. But if we wanted a
+specific version we can do: `http://localhost:8501/v1/models/vit/versions/1:predict`.
```
Comment (osanseviero) [inline tf-serving-vision.md:383]:
```suggestion
The REST API is -
`http://localhost:8501/v1/models/vit:predict` following the specification from
[here](https://www.tensorflow.org/tfx/serving/api_rest#predict_api). By default,
this always picks up the latest version of the model. But if you wanted a
specific version you can use: `http://localhost:8501/v1/models/vit/versions/1:predict`.
```

File: tf-serving-vision.md [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+
+# Querying the gRPC Endpoint
+
+While REST is quite popular in the API world many applications often
```
Comment (osanseviero) [inline tf-serving-vision.md:387]:
```suggestion
While REST is quite popular in the API world, many applications often
```

File: tf-serving-vision.md [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+benefit from gRPC. [This post](https://blog.dreamfactory.com/grpc-vs-rest-how-does-grpc-compare-with-traditional-rest-apis/)
+does a good job comparing the two ways of deployment. gRPC is usually
+preferred for low-latency, highly scalable, and distributed systems.
+
+There are a couple of steps are. First, we need to open a communication
```
Comment (osanseviero) [inline tf-serving-vision.md:392]:
```suggestion
There are a couple of steps are. First, you need to open a communication
```

File: tf-serving-vision.md [hunk 7]
```diff
@@ -0,0 +1,489 @@
...
+import grpc
+from tensorflow_serving.apis import predict_pb2
+from tensorflow_serving.apis import prediction_service_pb2_grpc
+
+
+channel = grpc.insecure_channel("localhost:8500")
+stub = prediction_service_pb2_grpc.PredictionServiceStub(channel)
+```
+
+Then, we create the request payload:
```
Comment (osanseviero) [inline tf-serving-vision.md:405]:
```suggestion
Then, create the request payload:
```

File: tf-serving-vision.md [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+```py
+request = predict_pb2.PredictRequest()
+request.model_spec.name = "vit"
+request.model_spec.signature_name = "serving_default"
+request.inputs[serving_input].CopyFrom(tf.make_tensor_proto([b64str]))
+```
+
+We can determine the `serving_input` key programmatically like so:
```
Comment (osanseviero) [inline tf-serving-vision.md:414]:
```suggestion
You can determine the `serving_input` key programmatically like so:
```

File: tf-serving-vision.md [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+```py
+loaded = tf.saved_model.load(f"{MODEL_DIR}/{VERSION}")
+serving_input = list(
+    loaded.signatures["serving_default"].structured_input_signature[1].keys()
+)[0]
+print("Serving function input:", serving_input)
+# Serving function input: string_input
+```
+
+Now, we can get some predictions:
```
Comment (osanseviero) [inline tf-serving-vision.md:425]:
```suggestion
Now, you can get some predictions:
```

File: tf-serving-vision.md [hunk 10]
```diff
@@ -0,0 +1,489 @@
...
+model_spec {
+  name: "resnet"
+  version {
+    value: 1
+  }
+  signature_name: "serving_default"
+}
+```
+
+We can also fetch the key-values of our interest from the above results like so:
```
Comment (osanseviero) [inline tf-serving-vision.md:466]:
```suggestion
You can also fetch the key-values of our interest from the above results like so:
```

File: tf-serving-vision.md [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+# ([b'Egyptian cat'], [0.8966591954231262])
+```
+
+# Wrapping Up
+
+In this post, we learned how to deploy a TensorFlow vision model from
+Transformers with TF Serving. While local deployments are great for
+weekend projects, we would want to be able to scale these deployments to
+serve many users. In the next series of posts, we'll see how to scale up
+these deployments with Kubernetes and Vertex AI.
```
Comment (osanseviero) [inline tf-serving-vision.md:481]:
```suggestion
In this post, we learned how to deploy a TensorFlow vision model from
Transformers with TF Serving. While local deployments are great for
weekend projects, we would want to be able to scale these deployments to
serve many users. In the next series of posts, you'll see how to scale up
these deployments with Kubernetes and Vertex AI.
```

**Review by osanseviero — APPROVED** [review]
This is an excellent blog post. The grammar, structure and redaction were really excellent, and the code is very clear to understand and follow. I left some suggestions but feel free to close them if you disagree, it's your blog post after all!

What do you think of deploying this on Monday? We have a couple of other blog posts this week already

File: tf-serving-vision.md [hunk 12]
```diff
@@ -0,0 +1,489 @@
...
+</a>
+
+In the past few months, the Hugging Face team and external contributors
+added a variety of vision models in TensorFlow to Transformers. This
+list is growing comprehensively and already includes state-of-the-art
+pre-trained models like [Vision Transformer](https://huggingface.co/docs/transformers/main/en/model_doc/vit),
+[Masked Autoencoders](https://huggingface.co/docs/transformers/model_doc/vit_mae),
+[RegNet](https://huggingface.co/docs/transformers/main/en/model_doc/regnet),
+[ConvNeXt](https://huggingface.co/docs/transformers/model_doc/convnext),
+and so on.
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tf-serving-vision.md:38]: There are more vision models in TF, which I didn't mention for brevity. What would you suggest in that case?

File: tf-serving-vision.md [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+      name: serving_default_pixel_values:0
+The given SavedModel SignatureDef contains the following output(s):
+  outputs['logits'] tensor_info:
+      dtype: DT_FLOAT
+      shape: (-1, 1000)
+      name: StatefulPartitionedCall:0
+Method name is: tensorflow/serving/predict
+```
+
+As can be noticed the model accepts single 4-d inputs (namely
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tf-serving-vision.md:99]: I intentionally kept it to `notice` for being more inclusive as not everyone can see :) But okay to change to `see` if that's preferred.

File: tf-serving-vision.md [hunk 14]
```diff
@@ -0,0 +1,489 @@
...
+the Colab Notebook shown at the beginning.
+
+# Saving the Model
+
+All TensorFlow models in 🤗 Transformers have a method named
+`save_pretrained()`. With it, we can serialize the model weights in
+the h5 format as well as in the standalone [SavedModel format](https://www.tensorflow.org/guide/saved_model).
+TF Serving needs a model to be present in the SavedModel format. So, let's first
+load a Vision Transformer model and save it:
+
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tf-serving-vision.md:62]: Sure.

File: tf-serving-vision.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (osanseviero) [reply tf-serving-vision.md:38]: maybe ", and many others!" WDYT?

File: tf-serving-vision.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (osanseviero) [reply tf-serving-vision.md:99]: Actually you're totally right, my fault! Let's keep the noticed!

File: tf-serving-vision.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply tf-serving-vision.md:38]: Yes, sure!

File: tf-serving-vision.md (hunk 1 above)
Comment (sayakpaul) [reply tf-serving-vision.md:2]: @Reeray I am happy to follow any instructions too. Appreciate the help in advance.

**Review by osanseviero — APPROVED** [review]
Looks great! Let's deploy this! 🔥

**Pull request merged**